packages/peregrine/lib/store/actions/cart/asyncActions.js

```diff
@@ -72,22 +64,14 @@ export const addItemToCart = (payload = {}) => {
                 sku: item.sku
             };
 
-            await addItemMutation({
+            const { data } = await addItemMutation({
                 variables
             });
 
-            // 2019-02-07  Moved these dispatches to the success clause of
-            // addItemToCart. The cart should only open on success.
-            // In the catch clause, this action creator calls its own thunk,
-            // so a successful retry will wind up here anyway.
-            await dispatch(
-                getCartDetails({
-                    fetchCartId,
-                    fetchCartDetails
-                })
-            );
-            await dispatch(toggleDrawer('cart'));
-            dispatch(actions.addItem.receive());
+            // The response contains the new updated cart state.
+            const { response } = data;
+            const { cart: details } = response;
+            dispatch(actions.addItem.receive({ details }));
```

Comment:
Instead of re-spreading the `details` after each mutation, we could try to only add/remove the relevant item in the cart. With that logic in place we could take advantage of [optimistic UI](https://www.apollographql.com/docs/angular/features/optimistic-ui/) which would allow us to update the UI without having to wait for GQL to respond. We could still display the error state if the server eventually responded with one, but we wouldn't be stuck behind the network request.